Resources/doc/default_configuration.rst
Default Configuration
=====================

.. code-block:: yml

    rps_contact:
        db_driver: orm
        contact_per_page: 25                                        # number of contacts to show on a page
        default_avatar: bundles/rpscontact/images/avatar.png        # relative path of the default avatar
        filter_image: true                                          # resize images

        class:
            model: RPS\ContactBundle\Entity\Entry                   # (optional) contact model class
            manager: RPS\ContactBundle\Entity\EntryManager          # (optional) contact manager class
            pager : RPS\CoreBundle\Pager\PagerfantaORM              # (optional) pager class

        view:
            list: RPSContactBundle:Contact:index.html.twig
            show: RPSContactBundle:Contact:show.html.twig
            new: RPSContactBundle:Contact:new.html.twig
            edit: RPSContactBundle:Contact:edit.html.twig

        form:
            contact:
                name: rps_contact_contact
                type: rps_contact_contact
                class: RPS\ContactBundle\Form\Type\ContactType      # contact form class

        service:
            pager: ~                                                # (optional) custom pager service

        image:
            avatar:                                                 # avatar image configuration
                quality: 100
                width: 150
                height: 150

            thumb:                                                  # thumb image configuration
                quality: 100
                width: 100
                height: 100


Each configuration option can be overridden in your app/config/config.yml file


Other topics
============

#. `Installation`_

#. `Doctrine Configuration`_

#. `Image Manager Configuration`_

#. `Pager Configuration`_

#. `Views/Templates`_

.. _Installation: Resources/doc/index.rst
.. _Doctrine Configuration: Resources/doc/doctrine.rst
.. _`Image Manager Configuration`: Resources/doc/image_manager.rst
.. _Pager Configuration: Resources/doc/pager.rst
.. _`Views/Templates`: Resources/doc/views.rst


DependencyInjection/Configuration.php
<?php

/*
 * This file is part of the RPSContactBundle package.
 *
 * (c) Yos Okusanya <yos.okusanya@gmail.com>
 *
 * This source file is subject to the MIT license that is bundled
 * with this source code in the file LICENSE.
 */

namespace RPS\ContactBundle\DependencyInjection;

use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

/**
 * This is the class that validates and merges configuration from your app/config files
 *
 * To learn more see {@link http://symfony.com/doc/current/cookbook/bundles/extension.html#cookbook-bundles-extension-config-class}
 */
class Configuration implements ConfigurationInterface
{
    /**
     * {@inheritDoc}
     */
   public function getConfigTreeBuilder()
    {
        $treeBuilder = new TreeBuilder();

        $treeBuilder->root('rps_contact')
            ->children()
                ->scalarNode('db_driver')->defaultValue('orm')->end()
                ->integerNode('contact_per_page')->min(1)->defaultValue(25)->end()
                ->booleanNode('filter_image')->defaultTrue()->end()
                ->scalarNode('default_avatar')->end()

                ->arrayNode('service')->addDefaultsIfNotSet()
                    ->children()
                        ->scalarNode('pager')->end()
                    ->end()
                ->end()

                ->arrayNode('class')->addDefaultsIfNotSet()
                    ->children()
                        ->scalarNode('pager')->end()
                        ->scalarNode('model')->cannotBeEmpty()->end()
                        ->scalarNode('manager')->cannotBeEmpty()->end()
                    ->end()
                ->end()

                ->arrayNode('view')->addDefaultsIfNotSet()
                    ->children()
                        ->scalarNode('list')
                            ->cannotBeEmpty()
                            ->defaultValue('RPSContactBundle:Default:index.html.twig')
                        ->end()
                        ->scalarNode('show')
                            ->cannotBeEmpty()
                            ->defaultValue('RPSContactBundle:Default:show.html.twig')
                        ->end()
                        ->scalarNode('new')
                            ->cannotBeEmpty()
                            ->defaultValue('RPSContactBundle:Default:new.html.twig')
                        ->end()
                        ->scalarNode('edit')
                            ->cannotBeEmpty()
                            ->defaultValue('RPSContactBundle:Default:edit.html.twig')
                        ->end()
                    ->end()
                ->end()

                ->arrayNode('form')->addDefaultsIfNotSet()
                    ->children()
                        ->arrayNode('contact')->addDefaultsIfNotSet()
                            ->children()
                                ->scalarNode('name')
                                    ->cannotBeEmpty()
                                    ->defaultValue('rps_contact_contact')
                                ->end()
                                ->scalarNode('type')
                                    ->cannotBeEmpty()
                                    ->defaultValue('rps_contact_contact')
                                ->end()
                                ->scalarNode('class')
                                    ->cannotBeEmpty()
                                    ->defaultValue('RPS\ContactBundle\Form\Type\ContactType')
                                ->end()
                            ->end()
                        ->end()
                    ->end()
                ->end()

                ->arrayNode('image')->addDefaultsIfNotSet()
                    ->children()
                        ->arrayNode('avatar')->addDefaultsIfNotSet()
                            ->children()
                                ->integerNode('quality')
                                    ->cannotBeEmpty()
                                    ->defaultValue(100)
                                 ->end()
                                ->integerNode('width')
                                    ->cannotbeEmpty()
                                    ->defaultValue(150)
                                ->end()
                                ->integerNode('height')
                                    ->cannotbeEmpty()
                                    ->defaultValue(150)
                                ->end()
                            ->end()
                        ->end()
                        ->arrayNode('thumb')->addDefaultsIfNotSet()
                            ->children()
                                ->integerNode('quality')
                                    ->cannotbeEmpty()
                                    ->defaultValue(100)
                                ->end()
                                ->integerNode('width')
                                    ->cannotbeEmpty()
                                    ->defaultValue(50)
                                ->end()
                                ->integerNode('height')
                                    ->cannotbeEmpty()
                                    ->defaultValue(50)
                                ->end()
                            ->end()
                        ->end()
                    ->end()
                ->end()
            ->end()
        ->end();

        return $treeBuilder;
    }
}
